### PR TITLE
OCI storage

### PR DESCRIPTION
**Pull request by raulmldv**
OCI storage

Implemented an OCI compliant storage for pull and inspect operations. This storage can be directly used by `linuxkit` as its own cache, preventing it to pull the images again. The structure of the storage directory is as follows:

```
storage_dir/
|________index.json        -> points to the manifest blob for each image
|________oci-layout
|            blobs/sha256/
|            |_______manifest_hash  -> points to config and layers
|            |_______config_hash
|            |_______layer_hash
```
The storage module supports store and fetching operations. The first fetch validates the files for the requested image and loads the config in the memory (the following fetches will use this config). Only one config is stored as we do not need to keep track of all the stored images when building an EIF.

When later adding the OCI archive feature, extracted image contents can be directly added to the storage and used by `nitro-cli` and `linuxkit` as well.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: enclave_build/src/cache.rs [hunk 1]
```diff
@@ -0,0 +1,220 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// S&PDX-License-Identifier: Apache-2.0
```
Comment (petreeftime) [inline enclave_build/src/cache.rs:2]: Typo: Should be SPDX

File: enclave_build/src/cache.rs [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    /// If an index file exists at the path, it loads the file's contents into the 'cached_images'
+    /// field. If not, a new empty index file is created at that path.
+    pub fn new(root_path: &Path) -> Result<Self> {
+        // Create all missing folders, if not already created
+        fs::create_dir_all(&root_path).map_err(EnclaveBuildError::CacheInitError)?;
+
+        // Try to open the index file and read the contents.
+        // If the file is missing, create it.
+        OpenOptions::new()
+            .write(true)
```
Comment (petreeftime) [inline enclave_build/src/cache.rs:112]: Why is read required here?

File: enclave_build/src/cache.rs [hunk 3]
```diff
@@ -242,6 +242,145 @@ impl CacheManager {
...
+            // The 'cached_layers' hashmap will contain all layer files found in the cache for the current image
+            .for_each(|(name, file)| {
+                if let Ok(file_name) = name.into_string() {
+                    cached_blobs.insert(file_name, file);
+                }
+            });
+
+        // Iterate through each layer found in the image manifest and validate that it is stored in
+        // the cache by checking the digest
+        for layer_obj in layers_vec {
```
Comment (petreeftime) [inline enclave_build/src/cache.rs:330]: This should be split into a separate function to keep this function simple and then you can also use a `map` or `for_each`.

File: enclave_build/src/cache.rs [hunk 4]
```diff
@@ -242,6 +242,145 @@ impl CacheManager {
...
+                })?
+                .to_string();
+
+            // Get the cached layer file matching the digest
+            // If not present, then a layer file is missing, so return Error
+            let mut layer_file = cached_blobs.get(&layer_digest).ok_or_else(|| {
+                EnclaveBuildError::CacheMissError("Layer missing from cache.".to_string())
+            })?;
+            let mut layer_bytes = Vec::new();
+            layer_file.read_to_end(&mut layer_bytes).map_err(|_| {
```
Comment (petreeftime) [inline enclave_build/src/cache.rs:355]: Would `read_to_string` make more sense here?

File: enclave_build/src/cache.rs [hunk 5]
```diff
@@ -467,6 +470,39 @@ impl CacheManager {
...
+    ///
+    /// If the data is not correctly cached or a file is missing, it returns an error.
+    ///
+    /// If the image is not cached, it does not attempt to pull the image from remote.
+    pub fn fetch_image_details(&self, image_name: &str) -> Result<ImageDetails> {
+        let config_hash = self
+            .get_image_hash_from_name(&Self::normalize_reference(image_name))
+            .ok_or_else(|| EnclaveBuildError::CacheMissError("Image hash missing".to_string()))?;
+
+        // // Add algorithm prefix to the hash
```
Comment (petreeftime) [inline enclave_build/src/cache.rs:484]: nit: Double `//`.

**Review by petreeftime — COMMENTED** [review]
In general the code looks correct, however, I found it difficult to read the code. A lot of the comments say what the code is doing, but I would prefer to also have some comments about why it does this (ie. either include documentation about how this image cache looks like inline, or provide links to the specific documentation that is relevant).

File: enclave_build/src/cache.rs (hunk 2 above)
Comment (raulmldv) [reply enclave_build/src/cache.rs:112]: Missed it during clean up.

**Review by shtaked — COMMENTED** [review]
Generally looks good, the only thing, as I mentioned I would leverage `oci-distribution` structures for reading/storing the content of the cache.

File: enclave_build/src/cache.rs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+pub const CACHE_OCI_LAYOUT_FILE: &str = "oci-layout";
+
+/// Constants used for complying with the OCI cache structure
+pub const DEFAULT_MEDIA_TYPE: &str = "application/vnd.oci.image.manifest.v1+json";
+pub const REF_ANNOTATION: &str = "org.opencontainers.image.ref.name";
+pub const OCI_LAYOUT: (&str, &str) = ("imageLayoutVersion", "1.0.0");
+
+/// Manifest entry in index file
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct ManifestEntry {
```
[the remark this answers is not in the document]
Comment (raulmldv) [reply enclave_build/src/cache.rs:37]: I tried to match the `linuxkit` format of the index file, but I see that `oci-distirbution` has a more general structure which might work. This could also loosen the coupling with `linuxkit`.

File: enclave_build/src/cache.rs [hunk 7]
```diff
@@ -201,13 +300,110 @@ impl CacheManager {
...
     }
+
+    /// Add `docker.io` to references that are missing this registry so `linuxkit` can validate cache presence
+    fn normalize_reference(reference: &str) -> String {
+        let docker_prefix = "docker.io/";
+        if reference.starts_with(docker_prefix) {
+            return reference.to_string();
+        }
+
+        docker_prefix.to_owned() + reference
```
[the remark this answers is not in the document]
Comment (raulmldv) [reply enclave_build/src/cache.rs:490]: So `linuxkit` will take the image reference and automatically append `docker.io/` to search for it in the cache and will not recognize `public.ecr.aws/` as an already existing registry. If I don't save the image like this, `linuxkit` will search an image reference that starts with `docker.io/public.ecr.aws/` and will fail to find it in the cache and later to pull it.

File: enclave_build/src/cache.rs [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+impl CacheManager {
+    /// Creates a new CacheManager instance and returns it. As argument, a path to the root folder
+    /// of the cache should be provided.
+    ///
+    /// Apart from that, the function also creates (if not already created) all folders from the path
+    /// specified as argument.
+    ///
+    /// If an index file exists at the path, it loads the file's contents into the 'cached_images'
+    /// field. If not, a new empty index file is created at that path.
+    pub fn new(root_path: &Path) -> Result<Self> {
```
Comment (lauralt) [inline enclave_build/src/cache.rs:111]: Does it help to use `Path` instead of `PathBuf`? We would avoid all the `to_path_buf`s.

File: enclave_build/src/cache.rs [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        Ok(Self {
+            root_path: root_path.to_path_buf(),
+            cached_images,
+        })
+    }
+
+    fn fetch_index(root_path: &Path) -> Result<CacheIndex> {
+        match File::options()
+            .read(true)
+            .open(root_path.join(CACHE_INDEX_FILE_NAME))
```
Comment (lauralt) [inline enclave_build/src/cache.rs:410]: Can't we use `File::open` directly?

File: enclave_build/src/cache.rs [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+                .filter_map(|entry| {
+                    entry
+                        .annotations
+                        .get(REF_ANNOTATION)
+                        .map(|img_ref| (img_ref.to_string(), entry.digest.to_string()))
+                })
+                // Read manifest files and map digest to JSON content
+                .filter_map(|(img_ref, manifest_hash)| {
+                    Self::fetch_manifest(root_path, &manifest_hash)
+                        .ok()
```
Comment (lauralt) [inline enclave_build/src/cache.rs:139]: So if the `fetch_*` functions return an error, we discard it or what is happening here?

File: enclave_build/src/cache.rs [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+            .as_str()
+            .ok_or_else(|| {
+                EnclaveBuildError::CacheMissError(
+                    "Failed to get config digest from image manifest.".to_string(),
+                )
+            })?
+            .strip_prefix("sha256:")
+            .ok_or_else(|| {
+                EnclaveBuildError::CacheMissError(
+                    "Failed to get config digest from image manifest.".to_string(),
```
Comment (lauralt) [inline enclave_build/src/cache.rs:477]: Can we give a different error message here than the one above?

File: enclave_build/src/cache.rs [hunk 12]
```diff
@@ -143,6 +146,102 @@ impl CacheManager {
...
+        })?;
+        let config_digest = format!("{:x}", sha2::Sha256::digest(config_json.as_bytes()));
+
+        File::create(&blobs_path.join(&config_digest))
+            .map_err(EnclaveBuildError::CacheStoreError)?
+            .write_all(config_json.as_bytes())
+            .map_err(EnclaveBuildError::CacheStoreError)?;
+
+        // If index file present, read and append new image entry
+        let index_json: CacheIndex = match File::options().read(true).open(CACHE_INDEX_FILE_NAME) {
```
Comment (lauralt) [inline enclave_build/src/cache.rs:208]: Again I think you can use File::open, also I think you can make the `match` a bit more Rusty, by using `map_or_else` or something else.

File: enclave_build/src/cache.rs [hunk 13]
```diff
@@ -143,6 +146,102 @@ impl CacheManager {
...
+            Ok(file) => serde_json::from_reader(file).map_err(EnclaveBuildError::SerdeError)?,
+            Err(_) => CacheIndex::default(),
+        };
+        let mut index_content = index_json;
```
Comment (lauralt) [inline enclave_build/src/cache.rs:212]: Do we really need a new variable here?

File: enclave_build/src/cache.rs [hunk 14]
```diff
@@ -143,6 +146,102 @@ impl CacheManager {
         })
     }
 
+    /// Stores the image data provided as argument in the cache at the folder pointed
+    /// by the 'root_path' field.
+    pub fn store_image_data(&mut self, image_name: &str, image_data: &ImageData) -> Result<()> {
```
Comment (lauralt) [inline enclave_build/src/cache.rs:160]: This method is way too big, and complicated I would say? Can you think of some ways to simplify it, or maybe split it in multiple functions etc.?

**Review by lauralt — COMMENTED** [review]
First part of review.

File: enclave_build/src/cache.rs (hunk 13 above)
Comment (raulmldv) [reply enclave_build/src/cache.rs:212]: No, missed it.

File: enclave_build/src/cache.rs [hunk 15]
```diff
@@ -0,0 +1,764 @@
...
+/// {CACHE_ROOT_PATH}/blobs/sha256/hash_of_blob_2\
+/// etc.
+///
+/// A blob can be:
+///
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/manifest_hash - the image manifest, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/config_hash - the image configuration, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/layer_hash - one of the image layers, each in a separate gzip compressed tar file.
+#[derive(Clone)]
+pub struct CacheManager {
```
Comment (lauralt) [inline enclave_build/src/cache.rs:93]: I feel we have a design problem here. For example we are doing the same operations multiple times, e.g. instead of fetching the cache index file multiple times, we should do that just once, and keep it in the cache manager (as a separate field).
I'll try to think of other ways of how we can improve the cache manager abstraction.

**Review by shtaked — APPROVED** [review]

File: enclave_build/docs/cache.md [hunk 16]
```diff
@@ -0,0 +1,127 @@
+# Image Cache
+
+When building an enclave image using the Docker daemon, the image will be pulled and stored in the Docker cache. We want to implement the same functionality when choosing to build an OCI image. The resulted cache will follow the [OCI specification](https://github.com/opencontainers/image-spec), allowing tools like `linuxkit` to use it as its own cache, and also adding images from OCI archive with the same format.
```
Comment (lauralt) [inline enclave_build/docs/cache.md:3]: Please limit the lines in md files to 80 chars; this is the general recommendation, otherwise it is difficult to read such files in an IDE for example.

File: enclave_build/docs/cache.md [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+      "digest": "sha256:98806831de537d56f19b4b3cbf6ed80187374a1521a18d2d0c5f8a0e3962b2ae",
+      "annotations": {
+        "org.opencontainers.image.ref.name": "docker.io/test-oci:hello"
+      }
+    }
+  ]
+}
+```
+
+As we can see, the index holds one manifest entry, for one image stored. The number of manifests increases with each newly stored image. The manifest hash will point us to the manifest blob.
```
Comment (lauralt) [inline enclave_build/docs/cache.md:45]: As already mentioned, please avoid the use of `will`. Makes the doc hard to read.

File: enclave_build/docs/cache.md [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+### Get root folder
+
+So we can get access to the cache contents outside the `CacheManager` (in our case for `linuxkit`), we want to be able to return the location of the cache files. It will return the root folder of the cache, as decided by the values of the environment variables `XDG_DATA_HOME` or `HOME`.
+
+## Workflow
+
+To see how these operation fit with each other and the structure as well, let's explore the common workflow:
+
+1. Given an image name we want to build an EIF
+2. Check if the cache exists. If not, pull the image and store it to the new cache.
```
Comment (lauralt) [inline enclave_build/docs/cache.md:123]: Which new cache? Please be more explicit.

File: enclave_build/docs/cache.md [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+
+This operation will return the details of an image (image name, config digest and the config itself). These details are saved in the memory of the `CacheManager` and will hold the information needed to build the enclave image. This operation is called after validating the cache.
+
+### Get root folder
+
+So we can get access to the cache contents outside the `CacheManager` (in our case for `linuxkit`), we want to be able to return the location of the cache files. It will return the root folder of the cache, as decided by the values of the environment variables `XDG_DATA_HOME` or `HOME`.
+
+## Workflow
+
+To see how these operation fit with each other and the structure as well, let's explore the common workflow:
```
Comment (lauralt) [inline enclave_build/docs/cache.md:120]: operations*

File: enclave_build/docs/cache.md [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+
+When building our first image, the cache will probably not exist. This operation will create, prior to storing an image, an empty cache. It will contain the cache root folder, an index file, a blobs folder and the oci-layout.
+
+### Store
+
+This is the operation which adds an image to the cache. It will be called on an empty cache or after a cache miss when searching for the image. After pulling the image and extracting its details, its manifest entry will be added to the index and the blobs will be written, each to their file.
+
+### Validate
+
+After we find an image in the index or the in-memory cache, we want to validate its structure before using the data. The validation starts from the index file. It should contain our image's reference and the digest of the manifest. If fields are missing or the JSON structure does not comply, an error will be returned.
```
Comment (lauralt) [inline enclave_build/docs/cache.md:102]: Is the in-memory cache the same you are talking about until this point or are we discussing about two different caches?

File: enclave_build/docs/cache.md [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+So we can get access to the cache contents outside the `CacheManager` (in our case for `linuxkit`), we want to be able to return the location of the cache files. It will return the root folder of the cache, as decided by the values of the environment variables `XDG_DATA_HOME` or `HOME`.
+
+## Workflow
+
+To see how these operation fit with each other and the structure as well, let's explore the common workflow:
+
+1. Given an image name we want to build an EIF
+2. Check if the cache exists. If not, pull the image and store it to the new cache.
+3. Check if the image is in the cache (in-memory). If not, pull the image store add it to the cache.
+4. Validate the image's files in the cache. If it fails, pull the image and overwrite cache.
```
Comment (lauralt) [inline enclave_build/docs/cache.md:125]:
```suggestion
4. Validate the image's files in the cache. If the validation fails, pull the image and overwrite cache.
```

File: enclave_build/docs/cache.md [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+
+So we can get access to the cache contents outside the `CacheManager` (in our case for `linuxkit`), we want to be able to return the location of the cache files. It will return the root folder of the cache, as decided by the values of the environment variables `XDG_DATA_HOME` or `HOME`.
+
+## Workflow
+
+To see how these operation fit with each other and the structure as well, let's explore the common workflow:
+
+1. Given an image name we want to build an EIF
+2. Check if the cache exists. If not, pull the image and store it to the new cache.
+3. Check if the image is in the cache (in-memory). If not, pull the image store add it to the cache.
```
Comment (lauralt) [inline enclave_build/docs/cache.md:124]: The second sentence might be missing some words.

File: enclave_build/docs/cache.md (hunk 21 above)
Comment (lauralt) [reply enclave_build/docs/cache.md:125]: Overwrite cache with what? Do we remove the image or what are we doing?

File: enclave_build/docs/cache.md [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+
+Layers are serialized filesystem changes that go on top of each other, starting from the base filesystem we create the image with. The file will be binary content that, from Nitro CLI's point of view, just has to obey the content digest in the manifest.
+
+## Operations
+
+The operations offered by the `cache` module are as follows:
+
+### Create
+
+When building our first image, the cache will probably not exist. This operation will create, prior to storing an image, an empty cache. It will contain the cache root folder, an index file, a blobs folder and the oci-layout.
```
Comment (lauralt) [inline enclave_build/docs/cache.md:94]: The files in the empty cache will have a default, initial state, or are they going to be empty files? Please add details about this.

**Review by lauralt — COMMENTED** [review]
Thanks for adding documentation. Left some feedback for it.

File: enclave_build/docs/cache.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (raulmldv) [reply enclave_build/docs/cache.md:3]: As it is a simplified version of the specification (does not contain all the possible fields), I thought an example would make it easier to understand.

File: enclave_build/docs/cache.md (hunk 20 above)
Comment (raulmldv) [reply enclave_build/docs/cache.md:102]: The cache itself is made up of the file structure presented. The in-memory cache is what the `CacheManager` loads (the map: image reference -> config digest). I will clarify.

File: enclave_build/docs/cache.md (hunk 21 above)
Comment (raulmldv) [reply enclave_build/docs/cache.md:125]: Write the new content of the invalid files and truncate the file.

File: enclave_build/docs/cache.md (hunk 16 above)
Comment (petreeftime) [reply enclave_build/docs/cache.md:3]: I think rust-fmt is 100 or 120 chars/line and I think it's fine if we keep a limit that works on a modern PC. But the line needs to be split for sure.

Comment (lauralt) [reply enclave_build/docs/cache.md:3]: Yes, I don't have a strong preference for 80, mostly just for splitting the lines.

**Comment by raulmldv** [comment]
Refactored the implementation from a cache to a storage. Following our discussions, I came to the conclusion that we do not need to store config information in memory for all the images present in the index. We just need the one image used to build the EIF when running `enclave-build`. This behaviour does not resemble a cache, thus the renaming and documentation changes.

To avoid file reads for any storage access and loading information for images we do not use, a singe config is kept in memory and is loaded on the first fetch operation. The first fetch performs the files reads and validations and the next ones will just use the config to build the image details. Storing an image (after pulling from a remote registry) automatically loads the config without validations.

File: enclave_build/src/storage.rs [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+///
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/manifest_hash - the image manifest, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/config_hash - the image configuration, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/layer_hash - one of the image layers, each in a separate gzip compressed tar file.
+#[derive(Clone)]
+pub struct OciStorage {
+    /// The root folder of the storage
+    root_path: PathBuf,
+
+    /// The config of the OCI image used to build the EIF
```
Comment (lauralt) [inline enclave_build/src/storage.rs:64]: You should explain why this is optional.

File: enclave_build/src/storage.rs [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+    fs::{self},
+    path::{Path, PathBuf},
+};
+
+use oci_spec::image::ImageConfiguration;
+
+use crate::{EnclaveBuildError, Result};
+
+/// Root folder for the image storage.
+pub const STORAGE_ROOT_FOLDER: &str = "XDG_DATA_HOME";
```
Comment (lauralt) [inline enclave_build/src/storage.rs:28]: Do we need these as pub?

File: enclave_build/src/storage.rs [hunk 26]
```diff
@@ -59,4 +71,312 @@ impl OciStorage {
...
+    pub fn store_image_data(&mut self, image_name: &str, image_data: &ImageData) -> Result<()> {
+        // Create the folder where the image data will be stored. Each image blob will be stored in
+        // a file named by the SHA256 digest of the content.
+        let blobs_path = self.root_path.join(STORAGE_BLOBS_FOLDER);
+        fs::create_dir_all(&blobs_path).map_err(EnclaveBuildError::OciStorageStoreError)?;
+
+        // Each layer file will be named after the layer's digest hash
+        Self::store_layers(&blobs_path, image_data)?;
+
+        // Store the manifest
```
Comment (lauralt) [inline enclave_build/src/storage.rs:90]: This comment is not useful.

File: enclave_build/src/storage.rs [hunk 27]
```diff
@@ -59,4 +71,312 @@ impl OciStorage {
...
+
+        // Write oci_layout file from template constant
+        self.store_layout()?;
+
+        self.config = Some(deserialize_from_reader(config_content.as_bytes())?);
+
+        Ok(())
+    }
+
+    fn default_oci_index() -> OciImageIndex {
```
Comment (lauralt) [inline enclave_build/src/storage.rs:107]: Can't we implement Default instead?

File: enclave_build/src/storage.rs [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+/// {STORAGE_ROOT_PATH}/blobs/sha256/hash_of_blob_1\
+/// {STORAGE_ROOT_PATH}/blobs/sha256/hash_of_blob_2\
+/// etc.
+///
+/// A blob can be:
+///
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/manifest_hash - the image manifest, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/config_hash - the image configuration, stored as a JSON String.\
+/// {IMAGE_FOLDER_PATH}/blobs/sha256/layer_hash - one of the image layers, each in a separate gzip compressed tar file.
+#[derive(Clone)]
```
Comment (lauralt) [inline enclave_build/src/storage.rs:59]: Is the Clone needed?

File: enclave_build/src/lib.rs [hunk 29]
```diff
@@ -59,12 +60,12 @@ pub enum EnclaveBuildError {
...
     #[error("Error while parsing credentials: `{0}`")]
     CredentialsError(String),
-    #[error("Image cache initialization error: `{0:?}`")]
-    CacheInitError(std::io::Error),
-    #[error("Cache store operation failed: `{0:?}`")]
-    CacheStoreError(std::io::Error),
-    #[error("Cache entry not found or has wrong: `{0:?}`")]
-    CacheMissError(String),
+    #[error("Image storage initialization error: `{0:?}`")]
+    OciStorageInitError(std::io::Error),
```
Comment (lauralt) [inline enclave_build/src/lib.rs:64]: I've just noticed, you don't have to include `Error` in the variant name, it is clear that it's an error because it is part of an `Error` enum. Even if you only add other variants here, let's not propagate the issue, and name the new ones (or the changed ones) without the `Error` part.

File: enclave_build/src/storage.rs [hunk 30]
```diff
@@ -59,4 +71,312 @@ impl OciStorage {
...
+    use sha2::Digest;
+    use std::env::temp_dir;
+    use vmm_sys_util::rand;
+
+    /// This function stores the test image in a temporary directory and returns that directory and
+    /// the storage manager initialized with it as root path.
+    fn setup_temp_storage() -> (PathBuf, OciStorage) {
+        // Use a temporary dir as the storage root path
+        // Create temporary random path so tests running in parallel won't overlap
+        let root_dir = temp_dir().join(rand::rand_alphanumerics(5));
```
Comment (lauralt) [inline enclave_build/src/storage.rs:493]: isn't it easier to use `TempDir` from vmm-sys-util?

File: enclave_build/src/lib.rs (hunk 29 above)
Comment (raulmldv) [reply enclave_build/src/lib.rs:64]: I can remove `Error` for the ones that express a problem like `OciStorageNotFound` or `OciStorageMalformed`. The others feel weird without naming them failures or errors. This should be consistent with the naming throughout the repo.

File: enclave_build/src/storage.rs (hunk 25 above)
Comment (raulmldv) [reply enclave_build/src/storage.rs:28]: Might be used in the future as public, will remove it for now.

File: enclave_build/src/storage.rs (hunk 28 above)
Comment (raulmldv) [reply enclave_build/src/storage.rs:59]: I can remove it for now.

File: enclave_build/src/storage.rs (hunk 27 above)
Comment (raulmldv) [reply enclave_build/src/storage.rs:107]: This is the default index we use in this storage, not the default of the structure itself.

File: enclave_build/src/lib.rs (hunk 29 above)
Comment (lauralt) [reply enclave_build/src/lib.rs:64]: Hmm I don't necessarily find it weird, for example if you have `Error::OciStorageInit` you would read this as: we have an error in the oci storage initialization. This is one of the consequences of [namespacing](https://en.wikipedia.org/wiki/Namespace), i.e. the `Error` part is simply redundant. If the name sounds weird maybe it is missing some verb, but the Error part should not be the one that is missing.

Comment (raulmldv) [reply enclave_build/src/lib.rs:64]: Alright, makes sense. Changed the new ones.

File: enclave_build/src/storage.rs [hunk 31]
```diff
@@ -398,10 +398,38 @@ impl OciStorage {
...
+        }
+        let image_config = self.config.as_ref().ok_or(EnclaveBuildError::ConfigError)?;
+
+        // Add algorithm prefix to the hash
+        let image_hash = format!(
+            "sha256:{:x}",
+            sha2::Sha256::digest(
+                &serde_json::to_vec(&self.config).map_err(EnclaveBuildError::SerdeError)?
+            )
+        );
```
Comment (lauralt) [inline enclave_build/src/storage.rs:431]: I'm not sure if I suggested at some point to not have this as a separate function, probably I did that, sorry, but looking again how many times this is used, I think it would be better though to define one, so that in case we change the algorithm we won't need to update it in multiple places.

File: enclave_build/src/storage.rs [hunk 32]
```diff
@@ -398,10 +398,38 @@ impl OciStorage {
...
+    /// Fetches the image metadata from storage as an ImageDetails struct.
+    ///
+    /// If the data is not correctly stored or a file is missing, it returns an error.
+    ///
+    /// If the image is not stored, it does not attempt to pull the image from remote.
+    pub fn fetch_image_details(&mut self, image_name: &str) -> Result<ImageDetails> {
+        if self.config.is_none() {
+            self.config = Some(self.check_stored_image(image_name)?);
+        }
+        let image_config = self.config.as_ref().ok_or(EnclaveBuildError::ConfigError)?;
```
Comment (lauralt) [inline enclave_build/src/storage.rs:423]: I think you can have these in a single block by using `unwrap_or(_else?)`.

File: enclave_build/src/storage.rs [hunk 33]
```diff
@@ -567,4 +595,29 @@ pub mod tests {
 
         assert_eq!(config, expected_config);
     }
+
+    #[test]
+    fn test_get_image_hash_from_name() {
```
Comment (lauralt) [inline enclave_build/src/storage.rs:646]: is the test name correct?

File: enclave_build/src/storage.rs [hunk 34]
```diff
@@ -567,4 +595,29 @@ pub mod tests {
...
+        let (_root_path, mut storage) = setup_temp_storage();
+
+        let image_details = storage
+            .fetch_image_details(image::tests::TEST_IMAGE_NAME)
```
Comment (lauralt) [inline enclave_build/src/storage.rs:650]: Is self.config here None or Some? Can we test here the other scenario as well, e.g. if self.config is Some, the config is preserved?

Comment (raulmldv) [reply enclave_build/src/storage.rs:650]: I'm pretty sure that the config is `Some` after performing `setup_temp_storage` (does the store as well). Do we want to add a test with `None`?

Comment (lauralt) [reply enclave_build/src/storage.rs:650]: Ideally every code path should be tested, so if it's not too difficult, yes we should add a test for that.

File: enclave_build/src/storage.rs (hunk 32 above)
Comment (lauralt) [reply enclave_build/src/storage.rs]: Ah right, if it's too complicated we can leave it as it is.

File: enclave_build/docs/storage.md [hunk 35]
```diff
@@ -0,0 +1,171 @@
...
+|________oci-layout
+|            blobs/sha256/
+|            |_______98806831de537d56f19b4b3cbf6ed80187374a1521a18d2d0c5f8a0e3962b2ae -> manifest_hash points to config and layers
+|            |_______feb5d9fea6a5e9606aa995e879d862b825965ba48de054caab5ef356dc6b3412 -> config_hash
+|            |_______2db29710123e3e53a794f2694094b9b4338aa9ee5c40b930cb8063a1be392c54 -> layer_hash
+```
+
+### Examples
+
+To better understand the storage structure hand how the index and manifest point to other files, we
```
Comment (lauralt) [inline enclave_build/docs/storage.md:35]: hand -> and

File: enclave_build/docs/storage.md [hunk 36]
```diff
@@ -0,0 +1,171 @@
...
+increases with each newly stored image. The manifest hash points us to the manifest blob.
+
+#### Layout file
+
+oci-layout:
+```
+{"imageLayoutVersion":"1.0.0"}
+```
+
+Simply defines the OCI image layout version we are using.
```
Comment (lauralt) [inline enclave_build/docs/storage.md:67]: missing subject

File: enclave_build/docs/storage.md [hunk 37]
```diff
@@ -0,0 +1,171 @@
...
+manifest entry is added to the index and the blobs are written, each to their file.
+
+### Fetch details
+
+This operation returns the details of an image (image name, config digest and the config itself).
+These details are saved in the memory as the config of the image we are currently using to build
+the EIF. If a fetch was performed before, the config is already loaded and we can just return it.
+No other file read is needed.
+
+When we fetch an image without the config already loaded, we want to also that it is stored
```
Comment (lauralt) [inline enclave_build/docs/storage.md:131]: we want to also check?

File: enclave_build/docs/storage.md [hunk 38]
```diff
@@ -0,0 +1,171 @@
+# Image Local Storage
```
Comment (lauralt) [inline enclave_build/docs/storage.md:1]: The doc looks much better now, thanks for the updates!

**Review by lauralt — DISMISSED** [review]

File: about.toml [hunk 39]
```diff
@@ -4,11 +4,12 @@ accepted = [
     "Zlib",
```
Comment (lauralt) [inline about.toml:4]: the commit message needs some fixes (the signed offs are done twice).

File: enclave_build/Cargo.toml [hunk 40]
```diff
@@ -3,7 +3,7 @@ name = "enclave_build"
 version = "0.1.0"
 authors = ["The AWS Nitro Enclaves Team <aws-nitro-enclaves-devel@amazon.com>"]
 edition = "2018"
-rust-version = "1.58"
+rust-version = "1.59"
```
Comment (lauralt) [inline enclave_build/Cargo.toml:6]: Shouldn't we update this in all crates?

**Review by lauralt — APPROVED** [review]
